### PR TITLE
[#98] One style errors

### LINKED ISSUE #98: Use one style for showing parsing errors
# Clarification and motivation

Currently CLI uses different styles to signal syntax errors.

## Examples:
- **sort**
```
>>> coffer find --sort BAD_FORMAT
option --sort: Invalid sort format: "BAD_FORMAT".
Expected format is:
 * to sort by entry name: 'name:<direction>',
 * to sort by the entry's last modified date: 'date:<direction>',
 * to sort by a field's contents: '<fieldname>:contents:<direction>',
 * to sort by a field's last modified date: '<fieldname>:date:<direction>'.
Direction can be 'asc' or 'desc'.
Examples: 'name:desc', 'password:date:asc'.
...
```
- **filter**
```
>>> coffer find --filter BAD_FORMAT
option --filter: Invalid filter format: "BAD_FORMAT".
Expected format is:
 * to filter by entry name: 'name~<substring>',
 * to filter by entry's last modified date: 'date<op><date>',
 * to filter by a field's contents: '<fieldname>:contents~<substring>',
 * to filter by a field's last modified date: '<fieldname>:date<op><date>'.
<op> can be '<=', '>=', '<', '>', or '='.
<date> can be 'YYYY', 'YYYY-MM', 'YYYY-MM-DD', or 'YYYY-MM-DD HH:MM:SS'.
Examples: 'name~vault', 'date<2020-02', 'url:contents~google.com', 'pw:date<2020-02'.

Parser error:
1:11:
  |
1 | BAD_FORMAT
  |           ^
unexpected end of input
expecting ':' or fieldname
...
```
In case of `sort` custom messages are shown, while in `filter`, `megaparsec`'s error messages appear. To keep things consistent, we can use `megaparsec` for both.

# Acceptance criteria

Rewrite parsing of `--sort` option using `megaparsec` instead of `Text.splitOn`.

### PR DESCRIPTION
**Pull request by Kariel-Myrr**
## Description

Problem : Parse error messages in `find --sort` and `find --find` had different styles

Solution : rewrote `sort` parsing to Megaparsec instead of `splitOn(":")`

## Related issue(s)

Fixed #98 

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: lib/CLI/Parser.hs [hunk 1]
```diff
@@ -568,6 +549,41 @@ parseFieldNameWhile whileCond = do
 parseFieldContentsEof :: MParser FieldContents
 parseFieldContentsEof = FieldContents . T.pack <$> P.manyTill P.anySingle P.eof
 
+
```
Comment (DK318) [inline lib/CLI/Parser.hs:552]: Same

File: lib/CLI/Parser.hs [hunk 2]
```diff
@@ -568,6 +549,41 @@ parseFieldNameWhile whileCond = do
...
+  try parseSortMeansByNameOrDate
+  <|> parseSortMeansByFieldName
+
+parseSortMeansByFieldName :: MParser Sort
+parseSortMeansByFieldName = do
+  fieldName <- parseFieldNameWhile (/= ':')
+  void $ P.char ':'
+  P.choice (
+    [ (SortByFieldDate fieldName)     <$ (P.string "date")
+    , (SortByFieldContents fieldName) <$ (P.string "contents") ] :: [MParser Sort] )
```
Comment (DK318) [inline lib/CLI/Parser.hs:577]: You can avoid using parentheses like this
```suggestion
  P.choice @[] @_ @Sort $
    [ (SortByFieldDate fieldName)     <$ (P.string "date")
    , (SortByFieldContents fieldName) <$ (P.string "contents")
    ]
```
Btw, in list declarations we put a trailing `]` (see [code-style](https://github.com/serokell/coffer/blob/main/docs/code-style.md#list-declarations))

File: lib/CLI/Parser.hs [hunk 3]
```diff
@@ -568,6 +549,41 @@ parseFieldNameWhile whileCond = do
...
+
+
+parseSortMeansByNameOrDate :: MParser Sort
+parseSortMeansByNameOrDate = P.choice (
+  [ SortByEntryName <$ P.string "name"
+  , SortByEntryDate <$ (P.string "date") ] :: [MParser Sort] )
```
Comment (DK318) [inline lib/CLI/Parser.hs:583]: Same here

**Review by DK318 — CHANGES_REQUESTED** [review]

**Comment by DK318** [comment]
And, please, when you start merging it, clean up your `git` history

**Comment by DK318** [comment]
Please, don't resolve conversations on your own. They are resolving by the same person who opened it

File: lib/CLI/Parser.hs [hunk 4]
```diff
@@ -568,6 +546,39 @@ parseFieldNameWhile whileCond = do
...
 parseFieldContentsEof = FieldContents . T.pack <$> P.manyTill P.anySingle P.eof
 
+parseSort :: MParser (Sort, Direction)
+parseSort = do
+  sort <- parseSortMeans
+  void $ P.char ':'
+  direction <- parseSortDirection
+  return (sort, direction)
+
+
```
Comment (DK318) [inline lib/CLI/Parser.hs:556]: Another extra line :smile:

**Review by DK318 — APPROVED** [review]
LGTM overall

File: lib/CLI/Parser.hs [hunk 5]
```diff
@@ -568,6 +546,39 @@ parseFieldNameWhile whileCond = do
...
+parseSortDirection :: MParser Direction
+parseSortDirection =
+  try (P.string "asc" $> Asc)
```
Comment (sancho20021) [inline lib/CLI/Parser.hs:559]:
```suggestion
  P.string "asc" $> Asc
```
I predict that this `try` is not needed. Looks like we either need `try` before both alternatives or we don't need them at all.

**Review by sancho20021 — APPROVED** [review]
Ok!

File: lib/CLI/Parser.hs [hunk 6]
```diff
@@ -568,6 +546,39 @@ parseFieldNameWhile whileCond = do
...
+  <|> P.string "desc" $> Desc
+
+parseSortMeans :: MParser Sort
+parseSortMeans =
+  try parseSortMeansByNameOrDate
+  <|> parseSortMeansByFieldName
+
+parseSortMeansByFieldName :: MParser Sort
```
Comment (dcastro) [inline lib/CLI/Parser.hs:567]:
```suggestion
parseSortMeansByFieldContentsOrDate :: MParser Sort
```

File: lib/CLI/Parser.hs (hunk 5 above)
Comment (dcastro) [reply lib/CLI/Parser.hs:559]: Yeah I don't think it's needed either.

File: lib/CLI/Parser.hs [hunk 7]
```diff
@@ -568,6 +546,39 @@ parseFieldNameWhile whileCond = do
...
+
+
+parseSortDirection :: MParser Direction
+parseSortDirection =
+  try (P.string "asc" $> Asc)
+  <|> P.string "desc" $> Desc
+
+parseSortMeans :: MParser Sort
+parseSortMeans =
+  try parseSortMeansByNameOrDate
```
Comment (dcastro) [inline lib/CLI/Parser.hs:564]: We should have a test for checking that this `try` here backtracks as intended.

Essentially, we want to make sure that when a have a field named `name`, then this is interpreted correctly by the parser:

```
--sort name:contents:asc
       ^^^^
          | This `name` does not mean "the entry's name", it means "a field named 'name'"
```

I have a feeling this test is not going to pass though, we might have to rework a few things.

Comment (dcastro) [reply lib/CLI/Parser.hs:564]: Indeed, it won't pass:


```
$ coffer find / --sort name:contents:asc

option --sort: Invalid sort format: "name:contents:asc".
Expected format is:
 * to sort by entry name: 'name:<direction>',
 * to sort by the entry's last modified date: 'date:<direction>',
 * to sort by a field's contents: '<fieldname>:contents:<direction>',
 * to sort by a field's last modified date: '<fieldname>:date:<direction>'.
Direction can be 'asc' or 'desc'.
Examples: 'name:desc', 'password:date:asc'.

Parser error:
1:6:
  |
1 | name:contents:asc
  |      ^^^^
unexpected "cont"
expecting "asc" or "desc"
```

The `parseSortMeansByNameOrDate` parses succeeded in matching on `name`, it didn't backtrack, and then it proceeded to `parseSortDirection` and failed.

**Review by dcastro — CHANGES_REQUESTED** [review]

File: lib/CLI/Parser.hs (hunk 7 above)
Comment (Kariel-Myrr) [reply lib/CLI/Parser.hs:564]: I think it could be solved by changing the order of parsing means. So we first will parse field name -> contents/date, and if not we will parse name as keyword
But we will get err from `--sort name:contents` not for `name:direction` but for expecting `name:contents:-missing direction-`

Comment (Kariel-Myrr) [reply lib/CLI/Parser.hs:564]: Also we can prevent farther problems with amount of arguments by adding followedBy ':' in `parseSortMeansByFieldContentsOrDate` part. So it will be successfully ended if only we have another argument after

File: tests/golden/find-command/find-command.bats [hunk 8]
```diff
@@ -131,6 +131,20 @@ EOF
...
+  assert_output - <<EOF
+/
+  secrets/
+    a - [2000-01-01 01:01:01]
+      sortfield: c [2000-01-01 01:01:01]
+    b - [2000-01-01 01:01:01]
+      sortfield: b [2000-01-01 01:01:01]
+    c - [2000-01-01 01:01:01]
+      sortfield: a [2000-01-01 01:01:01]
+EOF
```
Comment (dcastro) [inline tests/golden/find-command/find-command.bats:147]: Nice :+1: 

Since this is testing a separate edge case, let's put this assertion in its own dedicated test, e.g. `test "sort on field with name 'date' and 'name'"`

And also add a similar test for a field named `date`

File: lib/CLI/Parser.hs (hunk 7 above)
Comment (dcastro) [reply lib/CLI/Parser.hs]: I'm not sure I understand. The error messages for `name:contents` and `name:contents:` seem reasonable to me:

```
$ coffer find / --sort name:contents 
...

Parser error:
1:14:
  |
1 | name:contents
  |              ^
unexpected end of input
expecting ':'
```

```
$ coffer find / --sort name:contents:
...

Parser error:
1:15:
  |
1 | name:contents:
  |               ^
unexpected end of input
expecting "asc" or "desc"
```

File: lib/CLI/Parser.hs [hunk 9]
```diff
@@ -11,9 +11,14 @@ module CLI.Parser
 
 import BackendName (BackendName, newBackendName)
 import CLI.Types
+  (Command(..), CopyOptions(CopyOptions), CreateOptions(CreateOptions),
+  DeleteFieldOptions(DeleteFieldOptions), DeleteOptions(DeleteOptions), Direction(..),
+  FieldInfo(FieldInfo), Filter(..), FilterField(FilterFieldByContents, FilterFieldByDate),
+  FilterOp(..), FindOptions(FindOptions), Options(Options), RenameOptions(RenameOptions),
+  SetFieldOptions(SetFieldOptions), SomeCommand(..), Sort(..), TagOptions(TagOptions),
+  ViewOptions(ViewOptions), parseFilterDate)
```
Comment (dcastro) [inline lib/CLI/Parser.hs:19]: No need for explicit import list here

**Review by dcastro — APPROVED** [review]
LGTM :+1: 

Just one last very small change, and then please clean up the commit history ^^

File: lib/CLI/Parser.hs (hunk 9 above)
Comment (Kariel-Myrr) [reply lib/CLI/Parser.hs:19]: But stylish-haskell made it :>

Comment (dcastro) [reply lib/CLI/Parser.hs:19]: Are you sure? AFAIK stylish-haskell doesn't turn implicit import lists into explicit.

I think you may have accidentally clicked the "code lens" that HLS adds above the import:

image

If you click that code lens (in grey), HLS will make the import list explicit.

Can you please revert this?

**Comment by dcastro** [comment]
Small typo in the first commit's message:

> So error diff betw one command

File: lib/CLI/Parser.hs (hunk 9 above)
Comment (DK318) [reply lib/CLI/Parser.hs:19]: We are using specific `stylish-haskell` version (0.13.0.0 AFAIR)
